embassy-stm32f4/src/rtc.rs

```diff
@@ -0,0 +1,290 @@
+                    diff = alarm_timestamp - now;
+                } else {
+                    diff = 0;
+                }
+
+                if diff < 5 {
+                    alarm.timestamp.set(u64::MAX);
+                    alarm.callback.get().map(|(f, ctx)| f(ctx));
+                } else if diff < arr as u64 {
+                    arr = diff as u32;
+                }
+            }
+
+            self.set_arr(arr);
+        });
+    }
+
+    unsafe fn on_interrupt(&mut self) {
+        // Clear interrupt flag
+        self.rtc.deref().sr.write(|w| w.uif().clear_bit());
+
+        self.reset_timestamp();
+        self.recompute();
+        // self.rtc.deref().cr1.modify(|_, w| w.cen().set_bit());
+    }
+
+    fn trigger_alarm(&self, n: usize, cs: &CriticalSection) {
+        // self.rtc.intenclr.write(|w| unsafe { w.bits(compare_n(n)) });
+
+        let alarm = &self.alarms.borrow(cs)[n];
+        alarm.timestamp.set(u64::MAX);
+
+        // Call after clearing alarm, so the callback can set another alarm.
+        alarm.callback.get().map(|(f, ctx)| f(ctx));
+    }
+
+    fn set_alarm_callback(&self, n: usize, callback: fn(*mut ()), ctx: *mut ()) {
+        interrupt::free(|cs| {
+            let alarm = &self.alarms.borrow(cs)[n];
+            alarm.callback.set(Some((callback, ctx)));
+        })
+    }
+
+    fn set_alarm(&self, n: usize, alarm_timestamp: u64) {
+        self.rtc.deref().cr1.modify(|_, w| w.cen().clear_bit());
+
+        interrupt::free(|cs| {
+            (&self.alarms.borrow(cs)[n]).timestamp.set(alarm_timestamp);
+        });
+
+        self.recompute();
+        self.rtc.deref().cr1.modify(|_, w| w.cen().set_bit());
+    }
+
+    pub fn alarm0(&'static self) -> Alarm<T> {
+        Alarm { n: 0, rtc: self }
+    }
+    pub fn alarm1(&'static self) -> Alarm<T> {
+        Alarm { n: 1, rtc: self }
+    }
+    pub fn alarm2(&'static self) -> Alarm<T> {
+        Alarm { n: 2, rtc: self }
+    }
+}
+
+impl<T: Instance> embassy::time::Clock for RTC<T> {
+    fn now(&self) -> u64 {
+        self.timestamp + self.rtc.deref().cnt.read().bits() as u64
+    }
+}
+
+pub struct Alarm<T: Instance> {
+    n: usize,
+    rtc: &'static RTC<T>,
+}
+
+impl<T: Instance> embassy::time::Alarm for Alarm<T> {
+    fn set_callback(&self, callback: fn(*mut ()), ctx: *mut ()) {
+        self.rtc.set_alarm_callback(self.n, callback, ctx);
+    }
+
+    fn set(&self, timestamp: u64) {
+        self.rtc.set_alarm(self.n, timestamp);
+    }
+
+    fn clear(&self) {
+        self.rtc.set_alarm(self.n, u64::MAX);
+    }
+}
+
+mod sealed {
+    pub trait Instance {}
+
+    impl Instance for crate::pac::TIM7 {}
+    impl Instance for crate::pac::TIM2 {}
+}
+
+/// Implemented by all RTC instances.
+pub trait Instance: sealed::Instance + Sized + 'static {
+    /// The interrupt associated with this RTC instance.
+    type Interrupt: OwnedInterrupt;
+
+    fn deref(&self) -> &tim6::RegisterBlock;
+    fn ptr() -> *const tim6::RegisterBlock;
```

Comment:
the stm32f401 doesn't have a tim6 module

Reply:
stm32f4 only currently supports stm32f405 in any case. I will add support for all MCUs with a future PR.